PowerShellTools.HostService/ServiceManagement/Debugging/PowershellDebuggingServiceAttachValidator.cs

```diff
@@ -0,0 +1,119 @@
+                        return Resources.ProcessAttachFailErrorBody;
+                    }
+                }
+            }
+            else
+            {
+                // if remote attaching, make sure that we are still in a remote session after entering the host
+                DebugScenario scenario = _debuggingService.GetDebugScenario();
+                if (scenario != DebugScenario.RemoteSession)
+                {
+                    ServiceCommon.Log("Failed to attach to remote process; scenario after invoke: {0}", scenario);
+                    return Resources.ProcessAttachFailErrorBody;
+                }
+            }
+            return string.Empty;
+        }
+
+        /// <summary>
+        /// Used to verify detachment from a runspace
+        /// </summary>
+        /// <param name="preScenario">The debug scenario before invoking Exit-PSHostProcess</param>
+        /// <returns>True if the detachment was verified, false otherwise</returns>
+        public bool VerifyDetachFromRunspace(DebugScenario preScenario, AutoResetEvent attachSemaphore)
+        {
+            // wait for invoke to finish swapping the runspaces if detaching from a local process
+            if (preScenario == DebugScenario.LocalAttach)
+            {
+                bool didTimeout = !(attachSemaphore.WaitOne(DebugEngineConstants.AttachRequestEventTimeout));
+                if (didTimeout)
+                {
+                    // if semaphore times out, check to see if runspace looks ok, if it does then we will move forward
+                    if (_debuggingService.GetDebugScenario() != DebugScenario.Local)
+                    {
+                        ServiceCommon.Log("Failed to detach from local process. Semaphore timed out and runspace is confirmed to not be local.");
+                        return false;
+                    }
+                }
+            }
+            else
+            {
+                // if remote attaching, make sure that we are still in a remote session after exiting the host
+                DebugScenario scenario = _debuggingService.GetDebugScenario();
+                if (scenario != DebugScenario.RemoteSession)
+                {
+                    ServiceCommon.Log(string.Format("Failed to detach from remote process; scenario after invoke: {0}", scenario));
+                    return false;
+                }
+            }
+            return true;
+        }
+
+        /// <summary>
+        /// Verfies the first part of attaching to a remote runspace, the entering into a remote sessiom with the remote machine
+        /// </summary>
+        /// <returns>True if the debug scenario indicates a remote session, false otherwise</returns>
+        public bool VerifyAttachToRemoteRunspace()
+        {
+            return !(_debuggingService.GetDebugScenario() == DebugScenario.Local);
+        }
+
+        /// <summary>

```

Comment:
!VerifyFromAttach?  [General Q: Are both of these needed, or at least as public methods?]